cg/apps/downsample/downsample.py

```diff
@@ -0,0 +1,169 @@
+"""API that handles downsampling of samples."""
+import logging
+from pathlib import Path
+
+from cg.apps.downsample.utils import case_exists_in_statusdb, sample_exists_in_statusdb
+from cg.apps.housekeeper.hk import HousekeeperAPI
+from cg.constants import SequencingFileTag
+from cg.exc import DownsampleFailedError
+from cg.meta.meta import MetaAPI
+from cg.meta.workflow.downsample.downsample import DownsampleWorkflow
+from cg.models.cg_config import CGConfig
+from cg.models.downsample.downsample_data import DownsampleData
+from cg.store import Store
+from cg.store.models import Family, FamilySample, Sample
+from cg.utils.calculations import multiply_by_million
+
+LOG = logging.getLogger(__name__)
+
+
+class DownsampleAPI(MetaAPI):
+    def __init__(
+        self,
+        config: CGConfig,
+        sample_id: str,
+        number_of_reads: float,
+        case_id: str,
+        dry_run: bool = False,
+    ):
+        """Initialize the API."""
+        super().__init__(config)
+        self.config: CGConfig = config
+        self.status_db: Store = config.status_db
+        self.hk_api: HousekeeperAPI = config.housekeeper_api
+        self.sample_id: str = sample_id
+        self.number_of_reads: float = number_of_reads
+        self.case_id: str = case_id
+        self.dry_run: bool = dry_run
+        self.downsample_data: DownsampleData = self.get_meta_data()
+        LOG.info(f"Starting DownsampleAPI for sample {sample_id}.")
+
+    def get_meta_data(self) -> DownsampleData:
+        """Return the DownSampleData.
+        Raises:
+            DownsampleFailedError
+        """
+        try:
+            return DownsampleData(
+                status_db=self.status_db,
+                hk_api=self.housekeeper_api,
+                sample_id=self.sample_id,
+                number_of_reads=self.number_of_reads,
+                case_id=self.case_id,
+                out_dir=Path(self.config.downsample),
+            )
+        except Exception as error:
+            raise DownsampleFailedError(repr(error))
+
+    def add_downsampled_sample_entry_to_statusdb(self) -> Sample:
+        """Add a down sampled sample entry to StatusDB."""
+        downsampled_sample: Sample = self.downsample_data.downsampled_sample
+        if sample_exists_in_statusdb(
+            status_db=self.status_db, sample_id=downsampled_sample.internal_id
+        ):
+            raise ValueError(f"Sample {downsampled_sample.internal_id} already exists in StatusDB.")
+        LOG.info(
+            f"New downsampled sample created: {downsampled_sample.internal_id} from {downsampled_sample.from_sample}"
+            f"Application tag set to: {downsampled_sample.application_version.application.tag}"
+            f"Customer set to: {downsampled_sample.customer}"
+        )
+        if not self.dry_run:
+            self.status_db.session.add(downsampled_sample)
```

Comment:
Consider wrapping this in a try/except and rollback the transaction if an error occurs. I think we need a general pattern for this though and it should probably not be handled throughout the codebase. So disregard this 😅